src/Voron/Impl/Paging/CryptoPager.cs

```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+            var subKey = stackalloc byte[32];
+            fixed (byte* ctx = _context)
+            fixed (byte* mk = _masterKey)
+            {
+                if (Sodium.crypto_kdf_derive_from_key(subKey, 32, num, ctx, mk) != 0)
+                    throw new InvalidOperationException("Unable to generate derived key");
+
+                var dataSize = (ulong)GetNumberOfPages(page) * Constants.Storage.PageSize;
 
-            ByteString subKey;
-            using (allocator.Allocate(32, out subKey))
+                var npub = (byte*)page + PageHeader.NonceOffset;
+                if (*(long*)npub == 0)
+                    Sodium.randombytes_buf(npub, sizeof(long));
+                else
+                    *(long*)npub = *(long*)npub + 1;
+
+                ulong macLen = MacLen;
+                var rc = Sodium.crypto_aead_chacha20poly1305_encrypt_detached(
+                    destination + PageHeader.SizeOf,
+                    destination + PageHeader.MacOffset,
+                    &macLen,
+                    (byte*)page + PageHeader.SizeOf,
+                    dataSize - PageHeader.SizeOf,
+                    (byte*)page,
+                    (ulong)PageHeader.NonceOffset,
+                    null,
+                    npub,
+                    subKey
+                );
+                Debug.Assert(macLen == MacLen);
+
+                if (rc != 0)
+                    throw new InvalidOperationException($"Unable to encrypt page {num}, rc={rc}");
+            }
+        }
+
+        private void DecryptPage(PageHeader* page)
+        {
+            var num = page->PageNumber;
+
+            var destination = (byte*)page;
+            var subKey = stackalloc byte[32];
+            fixed (byte* ctx = _context)
+            fixed (byte* mk = _masterKey)
             {
-                fixed (byte* ctx = _context)
-                fixed (byte* mk = _masterKey)
+                if (Sodium.crypto_kdf_derive_from_key(subKey, 32, num, ctx, mk) != 0)
+                    throw new InvalidOperationException("Unable to generate derived key");
+
+                var dataSize = (ulong)GetNumberOfPages(page) * Constants.Storage.PageSize;
+                var rc = Sodium.crypto_aead_chacha20poly1305_decrypt_detached(
+                    destination + PageHeader.SizeOf,
+                    null,
+                    (byte*)page + PageHeader.SizeOf,
+                    dataSize - PageHeader.SizeOf,
+                    (byte*)page + PageHeader.MacOffset,
+                    (byte*)page,
+                    (ulong)PageHeader.NonceOffset,
+                    (byte*)page + PageHeader.NonceOffset,
+                    subKey
+                );
+                if (rc != 0)
+                    throw new InvalidOperationException($"Unable to decrypt page {num}, rc={rc}");
+            }
+        }
+
+        public override void Dispose()
+        {
+            if (Disposed)
+                return;
+
+            Disposed = true;
```

Comment:
Missing `GC.SuppressFinalizer` and do we have any unmanaged resource we manage that we can't let the inner one handle it?